xarray/coding/strings.py

```diff
@@ -0,0 +1,219 @@
+"""Coders for strings."""
+from __future__ import absolute_import, division, print_function
+
+from functools import partial
+
+import numpy as np
+
+from ..core import indexing
+from ..core.pycompat import bytes_type, dask_array_type, unicode_type
+from ..core.variable import Variable
+from .variables import (
+    VariableCoder, lazy_elemwise_func, pop_to,
+    safe_setitem, unpack_for_decoding, unpack_for_encoding)
+
+
+def create_vlen_dtype(element_type):
+    # based on h5py.special_dtype
+    return np.dtype('O', metadata={'element_type': element_type})
+
+
+def check_vlen_dtype(dtype):
+    if dtype.kind != 'O' or dtype.metadata is None:
+        return None
+    else:
+        return dtype.metadata.get('element_type')
+
+
+def is_unicode_dtype(dtype):
+    return dtype.kind == 'U' or check_vlen_dtype(dtype) == unicode_type
+
+
+def is_bytes_dtype(dtype):
+    return dtype.kind == 'S' or check_vlen_dtype(dtype) == bytes_type
+
+
+class EncodedStringCoder(VariableCoder):
+    """Transforms between unicode strings and fixed-width UTF-8 bytes."""
+
+    def __init__(self, allows_unicode=True):
+        self.allows_unicode = allows_unicode
+
+    def encode(self, variable, name=None):
+        dims, data, attrs, encoding = unpack_for_encoding(variable)
+
+        contains_unicode = is_unicode_dtype(data.dtype)
+        encode_as_char = 'dtype' in encoding and encoding['dtype'] == 'S1'
+
+        if contains_unicode and (encode_as_char or not self.allows_unicode):
+            if '_FillValue' in attrs:
+                raise NotImplementedError(
+                    'variable {!r} has a _FillValue specified, but '
+                    '_FillValue is yet supported on unicode strings: '
+                    'https://github.com/pydata/xarray/issues/1647'
+                    .format(name))
+
+            string_encoding = encoding.pop('_Encoding', 'utf-8')
+            safe_setitem(attrs, '_Encoding', string_encoding, name=name)
+            # TODO: figure out how to handle this in a lazy way with dask
+            data = encode_string_array(data, string_encoding)
```

Comment:
so does this evaluate eagerly?

Reply:
Indeed, that's the case currently. We could probably do *this step* lazily with dask, but when we call `ensure_fixed_length_bytes` below it's also going to evaluate this into memory.

The challenge is that to write fixed-width strings to disk we need to know how much space to allocate *before* we start writing the data.

Alternative options:
1. Try to figure out the amount of space to allocate first by checking the maximum string length in a separate pass (for dask arrays). This would involve doing the calculation twice, but is probably better than eager evaluation.
2. Use encoding of some sort to indicate the data-type ahead of time, e.g., `dtype='S10'` to indicate that the data should be converted first into a string array of size 10. This might even already work in some form, I'm not quite sure with how the encoding logic is currently layered.

Potentially we should implement both options, issuing a warning in case (1). I don't know how common dask arrays of strings are yet, but I imagine handling this well will be useful eventually.

Reply:
In my work, I personally have never encountered a string array big enough to require chunking.

I just wish there were some way to be more explicit about any potential eager evaluation. In some cases, the user would rather see an error than have a big dask array silently start computing, since the later can potentially crash your whole system.

Reply:
I agree with you about being explicit. I guess we could add a keyword argument to `to_netcdf()`, e.g., `allow_computing_object_arrays`, defaulting to `False`?

We currently have to load string arrays with `dtype=object` twice:
- Once to figure out if they hold bytes or unicode (i.e., the true dtype for each chunk).
- A second time to figure out the maximum string length, if necessary.

In each case, this could be done by mapping over the dask array converting each chunk into a scalar. This would be safe from a memory consumption perspective (vs. computing the entire arrays), because dask already stores a fixed amount of metadata per chunk.

Reply:
I want to leave this to a follow-up issue, since it will be a significant under-taking and is not a regression from current behavior: https://github.com/pydata/xarray/issues/2069